### ISSUE TITLE
Different tokenisation results between oniguruma and RE2

### ISSUE THREAD

**Issue by bzz** [src-d/enry#225]
Right now envy uses regex-based tokeniser (until #218 at least).

We have 2 drop-in replacement regex engines: default from Go RE2 and default from Ruby [oniguruma](https://github.com/kkos/oniguruma).

Recent improvements done with Go module migration #219 surfaced a new issue: it seems that tokeniser produces a bit different results, depending on which regex engine is used :/

More specifically, the token frequencies built from linguist samples are different and high-level code-generator test catch by comparing with a fixture (pre-generated with RE2) and fail on oniguruma profiles like this https://github.com/src-d/enry/pull/219#issuecomment-482525632

We need to find the exact reason and depending on it decide, if we want to support 2 versions of fixtures or change something so there is no difference in output.

This also potentially affects #194

**Comment by creachadair** [comment]
Often I have found the differences between engines on the same regexp boil down to small variations in the handling of greed (in particular: whether operators are greedy), zero-width assertions, anchoring (e.g., whether it is on by default), case-sensitivity, and other meta-properties of the execution model that aren't well-specified. Obviously I don't know what the causes are in this cases yet, but that's where I would start looking for cases.

**Comment by bzz** [comment]
Steps to reproduce

```
$ go test ./internal/code-generator/... -run Test_GeneratorTestSuite -testify.m TestGenerationFiles
ok  	gopkg.in/src-d/enry.v1/internal/code-generator/generator	35.537s

$ go test -tags oniguruma  ./internal/code-generator/... -run Test_GeneratorTestSuite -testify.m TestGenerationFiles
FAIL	gopkg.in/src-d/enry.v1/internal/code-generator/generator	34.583s
```

Both should produce the same results if results of tokenisation of all linguist samples are the same.;
Going to add extra tests to verify just that.

**Comment by bzz** [comment]
I localised the problem down to the difference in handling what seems to be [latin1 encoded file](https://github.com/github/linguist/commit/77e10e89801c2a50925758ca0ddfbb172fe6c475#diff-35160f7167f48375bc47b8f079ca8b59R74).

While tokenising `thå filling` encoded in latin1 and read as UTF-8
 
**RE2** gets
```
th
filling
�
```

while **Oniguruma** gets

```
th
illing
�
f
```

**Comment by creachadair** [comment]
Of the three, only Flex really seems to be doing anything reasonable here. RE2 is close—but the order of the tokens is weird. I have no idea what Oniguruma is doing there, but it seems obviously broken in at least two ways.

**Comment by bzz** [comment]
Yup. And if the content is decoded from _latin1_ and encoded to _utf8_ with `charmap.ISO8859_1.NewDecoder().Bytes(content)` the result is an expected one:


**RE2**
```
th
filling
å
```

**oniguruma**
```
th
filling
å
```

I know that Linguist does use ICU-based https://github.com/brianmario/charlock_holmes but am not sure yet if it's part of the tokenisation.

**Comment by creachadair** [comment]
Yeah, I think either we should normalize the encoding or find a way to treat the Unicode replacement character as part of the token, e.g., `xxx�yyy` ⇒ `xxx�yyy`, or use it as a separator and discard it, e.g. ⇒ `xxx` `�` `yyy` ⇒ `xxx` `yyy`. It seems like Linguist does the latter maybe.

**Comment by bzz** [comment]
True. And linguist with flex-based tokeniser does not have this issue, so no need to encoding detection there. Thank you for suggestions, let me think a bit more about that..

**Comment by bzz** [comment]
After digging deeper - ot seems that the offending tokenization rule is [`extractAndReplaceRegular(content)`](https://github.com/src-d/enry/blob/ae43e1a91fa26d4386f98d8a100329a186610871/internal/tokenizer/tokenize.go#L147).

The `extractRemainders` is then called on it's output and does `bytes.Fields(content) on  `" \xe5 "` which results in extra `�` in RE2 case.

**Comment by bzz** [comment]
Doing equivalent operation in Ruby results in `ArgumentError: invalid byte sequence in UTF-8`

```ruby
$irb

"th\xdd filling".scan(/[0-9A-Za-z_\.@#\/\*]+/)
```

**Comment by bzz** [comment]
Digging a little bit deeper with oniguruma's C API using awesome [examples/simple.c](https://github.com/kkos/oniguruma/blob/master/sample/simple.c), it starts to look like this may rather be a bug in [go-oniguruma `Regexp.findAll()`](https://github.com/src-d/go-oniguruma/blob/788303971318a98ac6d485136e6ae95d155c0876/regex.go#L213) implementation.

**Comment by bzz** [comment]
Even in C, oniguruma does consistently produce strange result _in UTF8 mode_ for the non-valid input bytes in UTF8, like above. Seems like a bug upstream.

As all regex for tokenization are *not* using any [Unicode character classes]( https://golang.org/pkg/regexp/syntax/#hdr-Syntax), all RE2-based matches are conducted in ASCII-only mode, while go-oniguruma has hardcoded UTF8.

For our use-case the fix would be to force Oniguruma also use ASCII mode and that indeed produces identical results even for non-valid bytes in UTF8.

I will submit a patch to cgo part of https://github.com/src-d/go-oniguruma to have an option to override [hardcoded UTF8](https://github.com/src-d/go-oniguruma/blob/788303971318a98ac6d485136e6ae95d155c0876/chelper.c#L20),  expose it in Go as `MustCompileWithEncoding` (similar to `MustCompileWithOption`)  and as soon as it's merged, move entry [regex/oniguruma.go](https://github.com/src-d/enry/blob/ae43e1a91fa26d4386f98d8a100329a186610871/regex/oniguruma.go#L12) to use it.

**Comment by bzz** [comment]
Fixed by #227
